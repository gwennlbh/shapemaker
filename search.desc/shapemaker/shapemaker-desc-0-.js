searchState.loadedDescShard("shapemaker", 0, "Angle, stored in degrees\nReturns the argument unchanged.\nCreates an angle given an amount, and what a full turn is …\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nget SVG coordinates of the cell’s center instead of its …\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCorners of the region’s outline Does <em>not</em> match …\nadds dx and dy to the end of the region (dx and dy are <em>not</em> …\nIterates all points except the ones specified in the <code>except</code>…\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\niterates from left to right then top to bottom (in a “…\nIterates all points outlining the region, in clockwise …\nresized is like enlarged, but transforms from the center, …\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nThe layers are in order of top to bottom: the first layer …\nputs this layer on bottom, and the others above, without …\nputs this layer on top, and the others below, without …\nre-order layers. The first layer in the list will be on …\nreturns a list of all unique filters used throughout the …\nCreate a new canvas. The layers are in order of top to …\nAn iterator over the variants of Color\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nDotted(color, diameter, spacing)\nHatches(color, angle, thickness_ratio, spacing)\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the effective region the layer occupies, by …\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nStruct can be rendered as a CSS ruleset (e.g. no selectors)\nWhen merging multiple SVGAttributesRenderable, this string …\nStruct can be rendered as attributes of a SVG element\nStruct can be rendered as a SVG element\nSets cx and cy\nSets x and y\nSets width and height\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nSets x and y\nSets x1, y1 and x2, y2\nSets x, y, width and height according to the region\nSets width and height\nmax amplitude of this stem\nin milliseconds\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nHow many instants (individual ms values) have occurred …\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nBuilt-in <code>Easing</code> implementations.\nArguments: animation progress (from 0.0 to 1.0), canvas, …\nAn easing function for customizing animations.\nAn animation that only manipulates a single layer. See …\nBuilt-in <code>Easing</code> implementations.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns a new easing function using <code>func</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns a new easing function using <code>easing</code>.\nExample\nAn <code>Easing</code> function that eases in using a curve that backs …\nAn <code>Easing</code> function that eases in using a curve that …\nAn <code>Easing</code> function that eases in using a curve resembling …\nAn <code>Easing</code> function that eases in using a cubic (x^3) curve\nAn <code>Easing</code> function that eases in using a curve that …\nAn <code>Easing</code> function that eases in using an expenential curve\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using a cubic …\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using an …\nAn <code>Easing</code> function that eases in and out using a quadradic …\nAn <code>Easing</code> function that eases in and out using a quartic …\nAn <code>Easing</code> function that eases in and out using a quintic …\nAn <code>Easing</code> function that eases in and out using a sine wave\nAn <code>Easing</code> function that eases in using a quadradic (x^2) …\nAn <code>Easing</code> function that eases in using a quartic (x^4) …\nAn <code>Easing</code> function that eases in using a quintic (x^5) …\nAn <code>Easing</code> function that eases in using a sine wave\nAn <code>Easing</code> function that eases out using a curve that backs …\nAn <code>Easing</code> function that eases out using a curve that …\nAn <code>Easing</code> function that eases out using a curve resembling …\nAn <code>Easing</code> function that eases out using a cubic (x^3) curve\nAn <code>Easing</code> function that eases out using a curve that …\nAn <code>Easing</code> function that eases out using an expenential …\nAn <code>Easing</code> function that eases out using a quadradic (x^2) …\nAn <code>Easing</code> function that eases out using a quartic (x^4) …\nAn <code>Easing</code> function that eases out using a quintic (x^5) …\nAn <code>Easing</code> function that eases out using a sine wave\nEases in using a curve that backs away initially\nEases in using a curve that bounces progressively closer …\nEases in using a curve resembling the top-left arc of a …\nEases in using a cubic (x^3) curve\nEases in using a curve that bounces around the start …\nEases in using an expenential curve\nEases in and out using a curve that backs away initially\nEases in and out using a curve that bounces progressively …\nEases in and out using a curve resembling the top-left arc …\nEases in and out using a cubic (x^3) curve\nEases in and out using a curve that bounces around the …\nEases in and out using an expenential curve\nEases in and out using a quadradic (x^2) curve\nEases in and out using a quartic (x^4) curve\nEases in and out using a quintic (x^5) curve\nEases in and out using a sine wave\nEases in using a quadradic (x^2) curve\nEases in using a quartic (x^4) curve\nEases in using a quintic (x^5) curve\nEases in using a sine wave\nAn <code>Easing</code> function that eases linearly\nEases linearly\nAn error returned from <code>StandardEasing::try_from</code> indicating …\nEases out using a curve that backs away initially\nEases out using a curve that bounces progressively closer …\nEases out using a curve resembling the top-left arc of a …\nEases out using a cubic (x^3) curve\nEases out using a curve that bounces around the start …\nEases out using an expenential curve\nEases out using a quadradic (x^2) curve\nEases out using a quartic (x^4) curve\nEases out using a quintic (x^5) curve\nEases out using a sine wave\nAn enumeration of all strandard easings provided.\nduration is in milliseconds Animates with ease-in-out …\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nduration is in milliseconds\nAn <code>Easing</code> function that eases in using a curve that backs …\nAn <code>Easing</code> function that eases in using a curve that …\nAn <code>Easing</code> function that eases in using a curve resembling …\nAn <code>Easing</code> function that eases in using a cubic (x^3) curve\nAn <code>Easing</code> function that eases in using a curve that …\nAn <code>Easing</code> function that eases in using an expenential curve\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using a cubic …\nAn <code>Easing</code> function that eases in and out using a curve …\nAn <code>Easing</code> function that eases in and out using an …\nAn <code>Easing</code> function that eases in and out using a quadradic …\nAn <code>Easing</code> function that eases in and out using a quartic …\nAn <code>Easing</code> function that eases in and out using a quintic …\nAn <code>Easing</code> function that eases in and out using a sine wave\nAn <code>Easing</code> function that eases in using a quadradic (x^2) …\nAn <code>Easing</code> function that eases in using a quartic (x^4) …\nAn <code>Easing</code> function that eases in using a quintic (x^5) …\nAn <code>Easing</code> function that eases in using a sine wave\nAn <code>Easing</code> function that eases out using a curve that backs …\nAn <code>Easing</code> function that eases out using a curve that …\nAn <code>Easing</code> function that eases out using a curve resembling …\nAn <code>Easing</code> function that eases out using a cubic (x^3) curve\nAn <code>Easing</code> function that eases out using a curve that …\nAn <code>Easing</code> function that eases out using an expenential …\nAn <code>Easing</code> function that eases out using a quadradic (x^2) …\nAn <code>Easing</code> function that eases out using a quartic (x^4) …\nAn <code>Easing</code> function that eases out using a quintic (x^5) …\nAn <code>Easing</code> function that eases out using a sine wave\nEases in using a curve that backs away initially\nEases in using a curve that bounces progressively closer …\nEases in using a curve resembling the top-left arc of a …\nEases in using a cubic (x^3) curve\nEases in using a curve that bounces around the start …\nEases in using an expenential curve\nEases in and out using a curve that backs away initially\nEases in and out using a curve that bounces progressively …\nEases in and out using a curve resembling the top-left arc …\nEases in and out using a cubic (x^3) curve\nEases in and out using a curve that bounces around the …\nEases in and out using an expenential curve\nEases in and out using a quadradic (x^2) curve\nEases in and out using a quartic (x^4) curve\nEases in and out using a quintic (x^5) curve\nEases in and out using a sine wave\nEases in using a quadradic (x^2) curve\nEases in using a quartic (x^4) curve\nEases in using a quintic (x^5) curve\nEases in using a sine wave\nAn <code>Easing</code> function that eases linearly\nEases linearly\nAn error returned from <code>StandardEasing::try_from</code> indicating …\nEases out using a curve that backs away initially\nEases out using a curve that bounces progressively closer …\nEases out using a curve resembling the top-left arc of a …\nEases out using a cubic (x^3) curve\nEases out using a curve that bounces around the start …\nEases out using an expenential curve\nEases out using a quadradic (x^2) curve\nEases out using a quartic (x^4) curve\nEases out using a quintic (x^5) curve\nEases out using a sine wave\nAn enumeration of all strandard easings provided.\nReturns a collection of every variant of this enum.\nEases in using a sine wave\nEases out using a sine wave\nEases in and out using a sine wave\nEases in using a quadradic (x^2) curve\nEases out using a quadradic (x^2) curve\nEases in and out using a quadradic (x^2) curve\nEases in using a cubic (x^3) curve\nEases out using a cubic (x^3) curve\nEases in and out using a cubic (x^3) curve\nEases in using a quartic (x^4) curve\nEases out using a quartic (x^4) curve\nEases in and out using a quartic (x^4) curve\nEases in using a quintic (x^5) curve\nEases out using a quintic (x^5) curve\nEases in and out using a quintic (x^5) curve\nEases in using an expenential curve\nEases out using an expenential curve\nEases in and out using an expenential curve\nEases in using a curve resembling the top-left arc of a …\nEases out using a curve resembling the top-left arc of a …\nEases in and out using a curve resembling the top-left arc …\nEases in using a curve that backs away initially\nEases out using a curve that backs away initially\nEases in and out using a curve that backs away initially\nEases in using a curve that bounces around the start …\nEases out using a curve that bounces around the start …\nEases in and out using a curve that bounces around the …\nEases in using a curve that bounces progressively closer …\nEases out using a curve that bounces progressively closer …\nEases in and out using a curve that bounces progressively …\nEases linearly\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nTells the rendering engine what to do with a frame\nWhat data is sent to the output by the rendering engine …\nRender this frame and stop rendering afterwards\nRender this frame as usual\nDon’t run hooks or anything on this frame\nDon’t render this frame and stop rendering\nSkip to the next frame, don’t render this one\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nArguments: canvas, context, previous rendered beat, …\nHooks that are triggered within a regular hook Used to …\nArguments: canvas, context, previous rendered beat\nArguments: canvas, context, current milliseconds timestamp\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nTriggers when a note starts on one of the stems in the …\nTriggers when a note stops on one of the stems in the …\nthreshold is a value between 0 and 1: current amplitude / …\nWhether the hook should be run only once\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nRender a single frame at the given frame number. Skip all …\nDuration of the video, without taking into account a …\nAdds the given scene and a hook that switches to it …\nAdds the given scene, and a hook that switches to it when …\nAdds hooks from the given scene to the video. Hooks will …\nThe VST3 plugin factory entry point.\nThe CLAP plugin’s entry point.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nAudio In signal name\nMaps automation parameter indices to their names\nReturns the argument unchanged.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nCalls <code>U::from(self)</code>.\nMIDI In signal name\nReturns a new probe with the <code>added_at</code> field set to the …\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.\nStore a data point.\nStore a audio data point.\nStore a automation data point.\nStore a midi data point.\nReturns the argument unchanged.\nCalls <code>U::from(self)</code>.")