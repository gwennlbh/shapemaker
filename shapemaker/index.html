<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `shapemaker` crate."><title>shapemaker - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-aa0817cf.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="shapemaker" data-themes="" data-resource-suffix="" data-rustdoc-version="1.90.0 (1159e78c4 2025-09-14)" data-channel="1.90.0" data-search-js="search-fa3e91e5.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-68b7e25d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-eebb9057.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../shapemaker/index.html">shapemaker</a><span class="version">1.2.2</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#reexports">Crate Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>shapemaker</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/shapemaker/lib.rs.html#1-55">Source</a> </span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">ยง</a></h2><dl class="item-table reexports"><dt id="reexport.Angle"><code>pub use geometry::<a class="struct" href="geometry/angle/struct.Angle.html" title="struct shapemaker::geometry::angle::Angle">Angle</a>;</code></dt><dt id="reexport.Axis"><code>pub use geometry::<a class="enum" href="geometry/axis/enum.Axis.html" title="enum shapemaker::geometry::axis::Axis">Axis</a>;</code></dt><dt id="reexport.Containable"><code>pub use geometry::<a class="trait" href="geometry/region/trait.Containable.html" title="trait shapemaker::geometry::region::Containable">Containable</a>;</code></dt><dt id="reexport.Point"><code>pub use geometry::<a class="struct" href="geometry/point/struct.Point.html" title="struct shapemaker::geometry::point::Point">Point</a>;</code></dt><dt id="reexport.Region"><code>pub use geometry::<a class="struct" href="geometry/region/struct.Region.html" title="struct shapemaker::geometry::region::Region">Region</a>;</code></dt><dt id="reexport.Canvas"><code>pub use graphics::<a class="struct" href="graphics/canvas/struct.Canvas.html" title="struct shapemaker::graphics::canvas::Canvas">Canvas</a>;</code></dt><dt id="reexport.Color"><code>pub use graphics::<a class="enum" href="graphics/color/enum.Color.html" title="enum shapemaker::graphics::color::Color">Color</a>;</code></dt><dt id="reexport.ColorMapping"><code>pub use graphics::<a class="struct" href="graphics/color/struct.ColorMapping.html" title="struct shapemaker::graphics::color::ColorMapping">ColorMapping</a>;</code></dt><dt id="reexport.ColoredObject"><code>pub use graphics::<a class="struct" href="graphics/objects/struct.ColoredObject.html" title="struct shapemaker::graphics::objects::ColoredObject">ColoredObject</a>;</code></dt><dt id="reexport.Fill"><code>pub use graphics::<a class="enum" href="graphics/fill/enum.Fill.html" title="enum shapemaker::graphics::fill::Fill">Fill</a>;</code></dt><dt id="reexport.FillOperations"><code>pub use graphics::<a class="trait" href="graphics/fill/trait.FillOperations.html" title="trait shapemaker::graphics::fill::FillOperations">FillOperations</a>;</code></dt><dt id="reexport.Filter"><code>pub use graphics::<a class="struct" href="graphics/filter/struct.Filter.html" title="struct shapemaker::graphics::filter::Filter">Filter</a>;</code></dt><dt id="reexport.FilterType"><code>pub use graphics::<a class="enum" href="graphics/filter/enum.FilterType.html" title="enum shapemaker::graphics::filter::FilterType">FilterType</a>;</code></dt><dt id="reexport.Layer"><code>pub use graphics::<a class="struct" href="graphics/layer/struct.Layer.html" title="struct shapemaker::graphics::layer::Layer">Layer</a>;</code></dt><dt id="reexport.LineSegment"><code>pub use graphics::<a class="enum" href="graphics/objects/enum.LineSegment.html" title="enum shapemaker::graphics::objects::LineSegment">LineSegment</a>;</code></dt><dt id="reexport.Object"><code>pub use graphics::<a class="enum" href="graphics/objects/enum.Object.html" title="enum shapemaker::graphics::objects::Object">Object</a>;</code></dt><dt id="reexport.ObjectSizes"><code>pub use graphics::<a class="struct" href="graphics/objects/struct.ObjectSizes.html" title="struct shapemaker::graphics::objects::ObjectSizes">ObjectSizes</a>;</code></dt><dt id="reexport.Transformation"><code>pub use graphics::<a class="enum" href="graphics/transform/enum.Transformation.html" title="enum shapemaker::graphics::transform::Transformation">Transformation</a>;</code></dt><dt id="reexport.CSSRenderable"><code>pub use rendering::<a class="trait" href="rendering/renderable/trait.CSSRenderable.html" title="trait shapemaker::rendering::renderable::CSSRenderable">CSSRenderable</a>;</code></dt><dt id="reexport.SVGAttributesRenderable"><code>pub use rendering::<a class="trait" href="rendering/renderable/trait.SVGAttributesRenderable.html" title="trait shapemaker::rendering::renderable::SVGAttributesRenderable">SVGAttributesRenderable</a>;</code></dt><dt id="reexport.SVGRenderable"><code>pub use rendering::<a class="trait" href="rendering/renderable/trait.SVGRenderable.html" title="trait shapemaker::rendering::renderable::SVGRenderable">SVGRenderable</a>;</code></dt><dt id="reexport.fonts"><code>pub use rendering::<a class="mod" href="rendering/fonts/index.html" title="mod shapemaker::rendering::fonts">fonts</a>;</code></dt><dt id="reexport.Animation"><code>pub use video::<a class="struct" href="video/animation/struct.Animation.html" title="struct shapemaker::video::animation::Animation">Animation</a>;</code></dt><dt id="reexport.AttachHooks"><code>pub use video::<a class="trait" href="video/hooks/trait.AttachHooks.html" title="trait shapemaker::video::hooks::AttachHooks">AttachHooks</a>;</code></dt><dt id="reexport.Scene"><code>pub use video::<a class="struct" href="video/scene/struct.Scene.html" title="struct shapemaker::video::scene::Scene">Scene</a>;</code></dt><dt id="reexport.Timestamp"><code>pub use video::<a class="struct" href="video/video/struct.Timestamp.html" title="struct shapemaker::video::video::Timestamp">Timestamp</a>;</code></dt><dt id="reexport.Video"><code>pub use video::<a class="struct" href="video/video/struct.Video.html" title="struct shapemaker::video::video::Video">Video</a>;</code></dt><dt id="reexport.animation"><code>pub use video::<a class="mod" href="video/animation/index.html" title="mod shapemaker::video::animation">animation</a>;</code></dt><dt id="reexport.context"><code>pub use video::<a class="mod" href="video/context/index.html" title="mod shapemaker::video::context">context</a>;</code></dt><dt><code>pub use graphics::<a class="enum" href="graphics/color/enum.Color.html" title="enum shapemaker::graphics::color::Color">Color</a>::*;</code></dt><dt><code>pub use graphics::<a class="enum" href="graphics/objects/enum.Object.html" title="enum shapemaker::graphics::objects::Object">Object</a>::*;</code></dt><dt><code>pub use synchronization::audio::<a class="enum" href="synchronization/audio/enum.MusicalDurationUnit.html" title="enum shapemaker::synchronization::audio::MusicalDurationUnit">MusicalDurationUnit</a>::*;</code></dt></dl><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="geometry/index.html" title="mod shapemaker::geometry">geometry</a></dt><dt><a class="mod" href="graphics/index.html" title="mod shapemaker::graphics">graphics</a></dt><dt><a class="mod" href="random/index.html" title="mod shapemaker::random">random</a></dt><dt><a class="mod" href="rendering/index.html" title="mod shapemaker::rendering">rendering</a></dt><dt><a class="mod" href="synchronization/index.html" title="mod shapemaker::synchronization">synchronization</a></dt><dt><a class="mod" href="ui/index.html" title="mod shapemaker::ui">ui</a></dt><dt><a class="mod" href="video/index.html" title="mod shapemaker::video">video</a></dt><dt><a class="mod" href="vst/index.html" title="mod shapemaker::vst">vst</a></dt></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="fn" href="fn.enabled_features.html" title="fn shapemaker::enabled_features">enabled_<wbr>features</a></dt></dl><script type="text/json" id="notable-traits-data">{"&[u8]":"<h3>Notable traits for <code>&amp;[<a class=\"primitive\" href=\"https://doc.rust-lang.org/1.90.0/std/primitive.u8.html\">u8</a>]</code></h3><pre><code><div class=\"where\">impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.90.0/std/io/trait.Read.html\" title=\"trait std::io::Read\">Read</a> for &amp;[<a class=\"primitive\" href=\"https://doc.rust-lang.org/1.90.0/std/primitive.u8.html\">u8</a>]</div>","&mut [u8]":"<h3>Notable traits for <code>&amp;mut [<a class=\"primitive\" href=\"https://doc.rust-lang.org/1.90.0/std/primitive.u8.html\">u8</a>]</code></h3><pre><code><div class=\"where\">impl <a class=\"trait\" href=\"https://doc.rust-lang.org/1.90.0/std/io/trait.Write.html\" title=\"trait std::io::Write\">Write</a> for &amp;mut [<a class=\"primitive\" href=\"https://doc.rust-lang.org/1.90.0/std/primitive.u8.html\">u8</a>]</div>"}</script></section></div></main></body></html>