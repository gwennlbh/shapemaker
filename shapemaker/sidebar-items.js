window.SIDEBAR_ITEMS = {"fn":["enabled_features"],"mod":["geometry","graphics","random","rendering","synchronization","ui","video","vst"]};