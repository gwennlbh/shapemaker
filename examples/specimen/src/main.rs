use shapemaker::*;

pub fn shapes_shed() -> Canvas {
    let mut canvas = Canvas::with_layers(vec![]);

    canvas.set_grid_size(3, 3);
    canvas.set_background(Color::White);

    canvas.layer("root").add_objects([
        Object::BigCircle(Point(0, 0)).colored(Color::Black),
        Object::CurveOutward(Point(1, 1), Point(2, 0), 5.0).colored(Color::Black),
        Object::CurveInward(Point(2, 1), Point(3, 0), 5.0).colored(Color::Black),
        Object::SmallCircle(Point(0, 1)).colored(Color::Black),
        Object::Line(Point(1, 1), Point(2, 2), 5.0).colored(Color::Black),
        Object::Rectangle(Point(0, 2), Point(0, 2)).colored(Color::Black),
        Object::Dot(Point(2, 3)).colored(Color::Black),
        Object::Polygon(
            Point(2, 1),
            vec![
                LineSegment::Straight(Point(3, 1)),
                LineSegment::Straight(Point(3, 2)),
            ],
        )
        .colored(Color::Black),
        Object::Text(Point(0, 0), "test".into(), 5.0).colored(Color::Black),
    ]);

    canvas
}

pub fn colors_shed() -> Canvas {
    let mut canvas = Canvas::with_layers(vec!["circles"]);
    canvas.set_grid_size(3, 3);
    canvas.canvas_outter_padding = 0;
    canvas.set_background(Color::White);

    let all_colors = vec![
        Color::Blue,
        Color::Cyan,
        Color::Yellow,
        Color::Orange,
        Color::Red,
        Color::Brown,
        Color::Purple,
        Color::Pink,
        Color::Green,
    ];

    for (color, point) in std::iter::zip(all_colors, canvas.world_region) {
        Object::Rectangle(point, point)
            .colored(color)
            .add_to(canvas.root());
    }

    canvas
}

pub fn grid() -> Canvas {
    let mut canvas = Canvas::with_layers(vec![]);
    canvas.set_grid_size(3, 3);
    canvas.set_background(Color::White);

    for point in canvas.world_region {
        Object::Dot(point)
            .colored(Color::Black)
            .add_to(canvas.root());
    }

    canvas
}

fn main() {
    grid()
        .render_to_svg_file("grid.svg")
        .expect("Failed to render grid");
    colors_shed()
        .render_to_svg_file("colorshed.svg")
        .expect("Failed to render colors_shed");
    shapes_shed()
        .render_to_svg_file("shapeshed.svg")
        .expect("Failed to render shapes_shed");
    shapes_shed()
        .render_to_png("shapeshed.png", 1000)
        .expect("Failed to render shapes_shed as PNG");
}

#[test]
fn test_grid() {
    use insta;
    insta::assert_snapshot! { grid().render_to_svg_string().unwrap() }
}

#[test]
fn test_colors_shed() {
    use insta;
    insta::assert_snapshot! { colors_shed().render_to_svg_string().unwrap() }
}

#[test]
fn test_shapes_shed() {
    use insta;
    insta::assert_snapshot! { shapes_shed().render_to_svg_string().unwrap() }
}
