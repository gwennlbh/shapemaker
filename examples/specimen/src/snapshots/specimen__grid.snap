---
source: examples/specimen/src/main.rs
expression: grid().render_to_svg_string().unwrap()
---
<svg height="170" viewBox="-10 -10 170 170" width="170" xmlns="http://www.w3.org/2000/svg">
<rect fill="white" height="170" width="170" x="-10" y="-10"/>
<g class="layer" data-layer="root">
<circle cx="0" cy="0" data-object="root--anon-0" r="2" style="fill: black;"/>
<circle cx="50" cy="0" data-object="root--anon-1" r="2" style="fill: black;"/>
<circle cx="100" cy="0" data-object="root--anon-2" r="2" style="fill: black;"/>
<circle cx="0" cy="50" data-object="root--anon-3" r="2" style="fill: black;"/>
<circle cx="50" cy="50" data-object="root--anon-4" r="2" style="fill: black;"/>
<circle cx="100" cy="50" data-object="root--anon-5" r="2" style="fill: black;"/>
<circle cx="0" cy="100" data-object="root--anon-6" r="2" style="fill: black;"/>
<circle cx="50" cy="100" data-object="root--anon-7" r="2" style="fill: black;"/>
<circle cx="100" cy="100" data-object="root--anon-8" r="2" style="fill: black;"/>
</g>
<defs/>
</svg>
